### PR TITLE
expedite job runs with backup query when build watch event goes missing

### PR DESCRIPTION
**Pull request by gabemontero**
quite possibly fixes https://github.com/openshift/origin/issues/16853

@openshift/devex fyi

also includes a format fix ... either use `?w=1` or focus on lines 172 to 210 of new version

### CODE REVIEW

**Comment by gabemontero** [comment]
[test]

**Comment by gabemontero** [comment]
@bparees - let me know, given the characteristics of the ext test flake and nature of this change, whether you want this in the 3.7 openshift jenkins image

thanks

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigWatcher.java [hunk 1]
```diff
@@ -123,248 +131,352 @@ public void doRun() {
...
+                        if (buildList.getItems().size() > 0) {
+                            logger.info("build backup query for "
+                                    + buildConfig.getMetadata().getName()
+                                    + " found new builds");
+                            BuildWatcher.onInitialBuilds(buildList);
+                        }
+                    }
+                };
+                Timer.get().schedule(backupBuildQuery, 10 * 1000,
+                        TimeUnit.MILLISECONDS);
```
Comment (bparees) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigWatcher.java]: Is there any risk of a race condition here if the build event comes in later?
ie:
1) buildconfig created
2) build created
3) this task is scheduled
4) build event shows up
5) race between 3+4 with potentially both launching corresponding jenkins jobs

or do we have sufficient locking around the creation of jenkins jobs (plus re-validating the build is still in a new state)?

**Comment by bparees** [comment]
if we can get some confidence around the race condition behavior i've asked about, i think it's ok to merge for 3.7.  But we definitely want to be sure this isn't going to lead to running 2 initial builds for a new buildconfig in some cases.

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigWatcher.java (hunk 1 above)
Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigWatcher.java]: Yep, good question.  Fortunately, I believe we have a good answer.

We have sufficient locking protection.  Ultimately it is no different then how the relist and event paths already work today.

Note this code:  https://github.com/openshift/jenkins-sync-plugin/blob/master/src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java#L301-L313

With the static method `addEventToJenkinsJobRun` synchronized, that is protection point 1.  It is single threaded for sure.  

Then, there is the check where if the build is no longer new, simply return.  As part of launching the jenkins job run, the build gets moved out of new state.

There is further protection later on at https://github.com/openshift/jenkins-sync-plugin/blob/master/src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java#L287-L300

If the job run corresponding to the build is already active, we punt there as well.

Comment (bparees) [reply src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigWatcher.java]: ok cool.

**Comment by bparees** [comment]
lgtm

**Comment by gabemontero** [comment]
thx 

[merge]

**Pull request merged**